src/widgets/wxWidgetsBasicUI.cpp

```diff
@@ -0,0 +1,236 @@
+/*!********************************************************************
+
+Audacity: A Digital Audio Editor
+
+@file wxWidgetsBasicUI.cpp
+
+Paul Licameli
+
+**********************************************************************/
+#include "wxWidgetsBasicUI.h"
+#include "MemoryX.h" // for Destroy_ptr
+#include "widgets/ErrorDialog.h"
+#ifdef HAS_SENTRY_REPORTING
+#include "widgets/ErrorReportDialog.h"
+#endif
+#include "widgets/AudacityMessageBox.h"
+#include "ProgressDialog.h"
+#include <wx/app.h>
+#include <wx/progdlg.h>
+#include <wx/windowptr.h>
+
+using namespace BasicUI;
+
+wxWidgetsWindowPlacement::~wxWidgetsWindowPlacement() = default;
+
+wxWidgetsBasicUI::~wxWidgetsBasicUI() = default;
+
+void wxWidgetsBasicUI::DoCallAfter(const Action &action)
+{
+   wxTheApp->CallAfter(action);
+}
+
+void wxWidgetsBasicUI::DoYield()
+{
+   wxTheApp->Yield();
+}
+
+namespace {
+wxWindow *GetParent(const BasicUI::WindowPlacement &placement)
+{
+   if (auto *pPlacement =
+       dynamic_cast<const wxWidgetsWindowPlacement*>(&placement))
+      return pPlacement->pWindow;
+   return nullptr;
+}
+}
+
+void wxWidgetsBasicUI::DoShowErrorDialog(
+   const BasicUI::WindowPlacement &placement,
+   const TranslatableString &dlogTitle,
+   const TranslatableString &message,
+   const ManualPageID &helpPage,
+   const BasicUI::ErrorDialogOptions &options)
+{
+   using namespace BasicUI;
+   bool modal = true;
+   auto parent = GetParent(placement);
+   switch (options.type) {
+      case ErrorDialogType::ModalErrorReport: {
+#ifdef HAS_SENTRY_REPORTING
+         ErrorReportDialog dlog(parent, dlogTitle, message, helpPage,
```

Comment:
Isn't it introduces dependency between BasicUI and error reporting module?

Reply:
No!  This makes wxWidgetsBasicUI dependent (at build time) on error reporting -- very true.  But BasicUI, which only contains the definition of the abstract base class Services, is not (build time) dependent on it.  And cutting that build-time dependency is exactly the point of all of this -- so that other things can depend on the lib-basic-ui but not on wxWidgets (other than wxBase utilities).

wxWidgetsBasicUI belongs to the Audacity program, and at application startup, Services are installed.  This is a "dependency injection."

But the hope is that libraries depending on lib-basic-ui might be re-used in other programs that supply a different, Qt-based implementation of Services at startup.  And those other libraries, then, will not need modifiction.

Reply:
And maybe that doesn't answer your original question?  But the code of wxWidgetsBasicUI is also written so it can compile and run whether HAS_SENTRY_REPORTING is defined, or not.

Reply:
Sorry, my mistake, ofcourse it's wxWidgetsBasicUI and not BasicUI